### PR TITLE
Make HTTP headers available as GUCs #800

### PR DESCRIPTION
**Pull request by ruslantalpa**
This PR makes request headers available as GUC vars for each request
It also sets `request.jwt.claims.role`, it useful when doing RLS through views.

Please look at runWithClaims from middleware if it needs some nicer style/formatting

### CODE REVIEW

**Comment by begriffs** [comment]
Thanks Ruslan, that's a long-requested feature. I'll review the code in a day or two.

**Comment by eldadj** [comment]
@ruslantalpa Thanks for this!

File: src/PostgREST/Middleware.hs [hunk 1]
```diff
@@ -48,6 +55,13 @@ runWithClaims conf eClaims app req =
       ]
       (toS $ "{\"message\":\""<>message<>"\"}")
 
+envVarToSql :: Text -> [(Text, Text)] -> [ByteString]
```
Comment (begriffs) [inline src/PostgREST/Middleware.hs]: Good idea to extract this into its own function. Maybe it makes more sense to put it in the QueryBuilder module, with return type `[SqlFragment]`?

Also `envVarToSql` sounds confusingly like unix environment variables. What about `setSqlVars` or something?

Adding a comment describing the motivation for the function would be nice, like how it's used for setting the role, jwt claims, and http request headers as GUCs.

File: src/PostgREST/Middleware.hs [hunk 2]
```diff
@@ -48,6 +55,13 @@ runWithClaims conf eClaims app req =
...
+envVarToSql prefix vars = varStmts
```
Comment (begriffs) [inline src/PostgREST/Middleware.hs]: This is minor, but you could inline `varStmts` as `map setVar vars` here. The extra name doesn't get us too much.

File: src/PostgREST/Auth.hs [hunk 3]
```diff
@@ -12,8 +12,8 @@ In the test suite there is an example of simple login function that can be used
 very simple authentication system inside the PostgreSQL database.
 -}
 module PostgREST.Auth (
-    claimsToSQL
-  , containsRole
+    -- claimsToSQL
```
Comment (begriffs) [inline src/PostgREST/Auth.hs]: You could remove this comment.

**Comment by ruslantalpa** [comment]
the build fail is due to some temporary network issue. Trigger rebuild in CircleCI and it should work

File: src/PostgREST/ApiRequest.hs [hunk 4]
```diff
@@ -92,6 +94,10 @@ data ApiRequest = ApiRequest {
   , iCanonicalQS :: ByteString
   -- | JSON Web Token
   , iJWT :: Text
+  -- | HTTP request headers
+  , iHeaders :: [(Text, Text)]
+  -- | Request Cookies
+  , iCookies :: Maybe [(Text, Text)]
```
Comment (begriffs) [inline src/PostgREST/ApiRequest.hs]: What's the difference in meaning here between `Nothing` and `Just []`?

Comment (ruslantalpa) [inline src/PostgREST/ApiRequest.hs]: No meaning, it was kind of an effect of the return type of the function used. Probably [] will be better

**Comment by begriffs** [comment]
Looks good! I made a note for myself to document this new behavior for the v0.4.1 docs

**Comment by pocman** [comment]
@begriffs Hi, I can't find the documentation for this :/
`current_setting(request.header.x-my-header)` point to an undefined_object.

**Comment by begriffs** [comment]
@pocman I added a little section about it to the docs: https://postgrest.com/en/v4.1/api.html#accessing-request-headers-cookies

In your case are you able to read some headers but not others, or are all headers coming back undefined?

**Comment by pocman** [comment]
Hi, it was just a typo 😉
Thanks. It's working fine.

**Pull request merged**